### DISCUSSION TITLE
WaitSets: What can I do to reduce my CPU usage?

### DISCUSSION THREAD

**Discussion by mahesh2392** [OpenDDS/OpenDDS#4207] [Q&A]
For high-rate data, the usage of a WaitSet may lead to a very high CPU usage if the WaitSet wakes up too often.

**In RTI DDS we can set WaitSet properties as below....**
DDS_WaitSetProperty_t prop = DDS_WaitSetProperty_t_INITIALIZER;
prop.max_event_count = 10;
prop.max_event_delay.sec = 30;
prop.max_event_delay.nanosec = 0;
  
DDSWaitSet* waitset = new DDSWaitSet(prop);

(long) max_event_count: Maximum number of trigger events to occur prior to waking up the WaitSet.
(DDS_Duration_t) max_event_delay: Maximum time delay from the first trigger event to waking up the WaitSet.

The properties configure the “wake up throttling” behavior of a WaitSet. Assuming there were no “active” conditions at the time the operation WaitSet::wait() was called, the WaitSet will wait until one of the three situations below occurs, whichever happens first:

The occurrence of max_event_count trigger events,
The passage of max_event_delay time from the occurrence of the first trigger event,
The passage of the maximum wait duration specified in the call to wait().

Like this do we can set waitset properties in openDDS so we can reduce CPU usage in case of high data rate?

**Comment by mitza-oci** [comment]
So "max_event_count" means "the minimum number of events needed to wake up the wait"?

It's not in the spec (current or proposed changes).  It could be added as an extension.

**Reply by mahesh2392** [reply]
So "max_event_count" means "the minimum number of events needed to wake up the wait"? -->Yes.
Is it possible to it as a extension ? or in my specific case is there any solution available in current openDDS implementation?

**Reply by mitza-oci** [reply]
Are you currently running this with RTI or just looking at their docs?  I'd be curious to know what the differences in performance are using standard DDS vs their proprietary extension.

I don't see a problem with adding something like this, but you would need to write the code or sponsor someone else to do so.

**Reply by mahesh2392** [reply]
I am just looking into RTI docs. Currently I am using openDDS but many of the listener in my application running on high data rate which cause huge CPU usage, I am looking into TIME_BASED_FILTER QOS but again as per docs "This QoS policy does
not conserve bandwidth as instance value changes are still sent to the subscriber process. It
only affects which samples are made available via the data reader." so I think it will wake up wait for each sample but I am not sure.

**Reply by mitza-oci** [reply]
Try the time based filter.  As it says, it's not going to change bandwidth (data sent on the network), but neither is the WaitSet change.

**Reply by jrw972** [reply]
Is TIME_BASED_FILTER like content filtering where the writer could filter if the implementation were improved?

**Reply by mitza-oci** [reply]
It is similar in some ways -- in the specs both kinds of filtering can avoid sending data on the network.  In OpenDDS, some cases of content filtering already do this optimization.

But the WaitSet proprietary extension from RTI is not related to QoS so that feature alone isn't going to change what's sent.

@mahesh2392 I suggest trying your use case with OpenDDS.  It's hard to know how something will perform without testing it.

**Reply by mahesh2392** [reply]
@mitza-oci Thanks for suggestion, with TIME_BASED_FILTER CPU usage reduced by almost 50%, but I cant use this because I want process every sample received to listener but in less CPU usage. So, is there any way instead of getting callback on_data_available() for every sample received can we just read values from reader as per our required periodicity?

**Reply by jrw972** [reply]
Yes.  Read/take at the desired interval.

**Reply by mahesh2392** [reply]
@jrw972 I try Read/take but still I see reader consume much CPU. I am not registering any listener to reader so no on_data_available() callback and just reading samples with some interval, so my question is even we are not attaching listener, every received event WaitSet wakes up and schedule the thread that is waiting on the WaitSet? is there any way to avoid this context switching as we are getting data on 10ms cycle rate.

**Reply by mahesh2392** [reply]
One more thing even publisher also taking much CPU usage. I am writing at 10ms periodicity.

**Reply by jrw972** [reply]
If you are reading at the same rate that the samples are coming in, then there is no advantage between a listener, waitset, or polling.  If you read at a lower rate to take advantage of batch processing, then polling will be more efficient because it avoids the overhead of the listener and waitset.  That is, you are processing multiple samples without context switching.  Getting back to your original question; the extra WaitSet properties from RTI are accomplishing the same idea of desensitizing the WaitSet so the caller would see more available samples per wakeup and therefore reduce overhead.

The question of CPU is relative.  It depends on message types, QoS, transport type, network conditions, OS configuration, the performance of application code, and the performance of OpenDDS code.  I suggest checking your configuration and profiling your applications to get a better understanding of their performance.
